tests/unit/lib/charms/sdcore_gnbsim/v0/test_fiveg_gnb_indentity_requirer.py

```diff
@@ -1,30 +1,42 @@
 # Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-import unittest
 from unittest.mock import call, patch
 
-from ops import testing
+import pytest
+from ops import BoundEvent, testing
 from test_charms.test_requirer_charm.src.charm import WhateverCharm  # type: ignore[import]
 
 TEST_CHARM_PATH = "charms.sdcore_gnbsim_k8s.v0.fiveg_gnb_identity.GnbIdentityRequirerCharmEvents"
+RELATION_NAME = "fiveg_gnb_identity"
 
+class TestGnbIdentityRequires:
 
-class TestGnbIdentityRequires(unittest.TestCase):
+    patcher_gnb_identity = patch(f"{TEST_CHARM_PATH}.fiveg_gnb_identity_available")
+
+    @pytest.fixture()
     def setUp(self) -> None:
+        self.mock_gnb_identity = TestGnbIdentityRequires.patcher_gnb_identity.start()
+        self.mock_gnb_identity.__class__ = BoundEvent
```

Comment:
Why do we need this? ain't there a nicer way?

Reply:
framework.observe(someobj.something_happened, self._on_something_happened)
The expected second parameter type is BoundEvent. When we mock it, the following error happens and setting the type to "BoundEvent" is required if we would like to check on emitted events.
```
Framework.observe requires a BoundEvent as second parameter, got <MagicMock name='fiveg_gnb_identity_available' id='140609246438992'>
```

Reply:
There tried an alternative approach without setting a type for mocked CharmEvents. In this case, we can not assert on calls for the mocked object but we could check the emitted events by using logs.  

```python
import logging
from unittest.mock import Mock, patch

import pytest
from ops import testing

from lib.charms.sdcore_nrf_k8s.v0.fiveg_nrf import NRFRequirerCharmEvents
from tests.unit.charms.sdcore_nrf.v0.dummy_requirer_charm.src.dummy_requirer_charm import (  # noqa: E501
    DummyFiveGNRFRequirerCharm,
)

DUMMY_REQUIRER_CHARM = "tests.unit.charms.sdcore_nrf.v0.dummy_requirer_charm.src.dummy_requirer_charm.DummyFiveGNRFRequirerCharm"  # noqa: E501
FIVEG_NRF_LIB = "lib.charms.sdcore_nrf_k8s.v0.fiveg_nrf.NRFRequirerCharmEvents"
RELATION_NAME = "fiveg_nrf"
REMOTE_APP_NAME = "dummy-nrf-requirer"
NRF_URL = "https://nrf.example.com"


class TestFiveGNRFRequirer:
    
    @pytest.fixture()
    def setUp(self) -> None:
        self.mock_nrf_requirer_charm_events = NRFRequirerCharmEvents()
        self.mock_nrf_requirer_charm_events.nrf_available = Mock(name='nrf_available')
        self.mock_nrf_requirer_charm_events.nrf_broken = Mock(name='nrf_broken')


    def tearDown(self) -> None:
        patch.stopall()

    @pytest.fixture(autouse=True)
    def harness(self, setUp, request):
        self.harness = testing.Harness(DummyFiveGNRFRequirerCharm)
        self.harness.begin()
        yield self.harness
        self.harness.cleanup()
        request.addfinalizer(self.tearDown)

    def _create_relation(self, remote_app_name: str):
        relation_id = self.harness.add_relation(
            relation_name=RELATION_NAME, remote_app=remote_app_name
        )
        self.harness.add_relation_unit(
            relation_id=relation_id, remote_unit_name=f"{remote_app_name}/0"
        )

        return relation_id

    def test_given_nrf_information_in_relation_data_when_relation_changed_then_nrf_available_event_emitted(  # noqa: E501
        self
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)

        relation_data = {
            "url": NRF_URL,
        }
        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )

    def test_given_nrf_information_not_in_relation_data_when_relation_changed_then_nrf_available_event_not_emitted(  # noqa: E501
        self
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)
        relation_data = {}

        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )
        # self.mock_fiveg_nrf_available.assert_not_called()
        self.mock_nrf_requirer_charm_events.nrf_available.assert_not_called()


    def test_given_invalid_nrf_information_in_relation_data_when_relation_changed_then_nrf_available_event_not_emitted(  # noqa: E501
        self
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)
        relation_data = {"pizza": "steak"}

        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )
        self.mock_nrf_requirer_charm_events.nrf_available.assert_not_called()

    def test_given_invalid_nrf_information_in_relation_data_when_relation_changed_then_error_is_logged(  # noqa: E501
        self, caplog
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)
        relation_data = {"pizza": "steak"}
        caplog.set_level(logging.DEBUG)
        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )
        assert "Invalid relation data: {'pizza': 'steak'}" in caplog.messages

    def test_given_nrf_information_in_relation_data_when_get_nrf_url_is_called_then_expected_url_is_returned(  # noqa: E501
        self,
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)

        relation_data = {
            "url": NRF_URL,
        }
        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )

        nrf_url = self.harness.charm.nrf_requirer.nrf_url
        assert nrf_url == NRF_URL

    def test_given_nrf_information_not_in_relation_data_when_get_nrf_url_then_returns_none(  # noqa: E501
        self, caplog
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)
        relation_data = {}

        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )
        caplog.set_level(logging.DEBUG)
        nrf_url = self.harness.charm.nrf_requirer.nrf_url
        assert nrf_url is None
        assert "Invalid relation data: {}" in caplog.messages

    def test_given_nrf_information_in_relation_data_is_not_valid_when_get_nrf_url_then_returns_none_and_error_is_logged(  # noqa: E501
        self, caplog
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)
        relation_data = {"pizza": "steak"}
        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )
        caplog.set_level(logging.DEBUG)
        nrf_url = self.harness.charm.nrf_requirer.nrf_url
        assert nrf_url is None
        assert "Invalid relation data: {'pizza': 'steak'}" in caplog.messages

    def test_given_nrf_relation_created_when_relation_broken_then_nrf_broken_event_emitted(
        self, caplog
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)
        caplog.set_level(logging.DEBUG)
        self.harness.remove_relation(relation_id)
        assert "Emitting custom event <NRFBrokenEvent via DummyFiveGNRFRequirerCharm/NRFRequires[fiveg_nrf]/on/nrf_broken[5]>." == caplog.messages[0]

```